### PR TITLE
Mark Assembly.CodeBase / Assembly.EscapedCodeBase as obsolete (#31127)

### PR DESCRIPTION
**Pull request by JoshSchreuder**

### CODE REVIEW

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:82]: I'm not sure if this is actually true now, can someone confirm?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:108]: Also not sure if this is true

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs
Comment (JoshSchreuder) [inline src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs:44]: Unsure if this should just be `Obsolete`d as well, or just pragma ignored.

File: docs/project/list-of-obsoletions.md
Comment (JoshSchreuder) [inline docs/project/list-of-obsoletions.md:18]: Needs a [Better Obsoletions](https://github.com/dotnet/designs/blob/master/accepted/2020/better-obsoletion/better-obsoletion.md) ID

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:88]: This should be just `string fileName = asm.Location;`; and the file:// comment should be deleted.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:84]: This check should replaced by asm.Location returning empty filename.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (vitek-karas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:88]: This also needs a check for `Assembly.Location` returning empty string:
* empty string is returned by assemblies loaded from memory - and will also be used by assemblies in a single file bundle.
* The code below will throw if this happens

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (vitek-karas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:84]: * empty string is returned by assemblies loaded from memory - and will also be used by assemblies in a single file bundle.
* the code below will throw if this happens

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (vitek-karas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:108]: This also needs a check for Assembly.Location returning empty string - same reasoning as above.

File: docs/project/list-of-obsoletions.md
Comment (jeffhandley) [reply docs/project/list-of-obsoletions.md:18]: It'll be `SYSLIB0012`. See https://github.com/dotnet/designs/pull/139 for more info if desired. And note that this PR will need to be rebased on top of https://github.com/dotnet/runtime/pull/39269 once that is merged.

File: src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/Hosting/AssemblyCatalogTests.cs
Comment (jeffhandley) [inline src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/Hosting/AssemblyCatalogTests.cs:39]: Once you apply `SYSLIB0012` as your `DiagnosticId`, you can change all of these to:
```suggestion
#pragma warning disable SYSLIB0012
```

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs
Comment (jeffhandley) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs:44]: You can obsolete this as well and use the same DiagnosticId. If you check out https://github.com/dotnet/runtime/pull/39269, you'll see that we're using a pattern of an `Obsoletions.cs` file to make the descriptions, ids, and the urlformat all reusable in the `src` files.

But you'll want to rebase on top of that PR once it's merged.

Comment (jeffhandley) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs:44]: Once #39269 is merged (very soon), there will be a common `Obsoletions.cs` file now even, where all obsoletion messages/diagnosticid values are maintained, and that file can be linked into all projects that use new obsoletions.

Comment (JoshSchreuder) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs:44]: Thanks @jeffhandley, I'll rebase and make those changes once that PR is merged

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs]: Updated based on feedback

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs]: Updated based on feedback

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs]: Updated based on feedback

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (jkotas) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:7559]: Asked question about this in https://github.com/dotnet/runtime/issues/31127#issuecomment-660108692

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs
Comment (JoshSchreuder) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs]: Updated to use common Obsoletions class

File: docs/project/list-of-obsoletions.md
Comment (JoshSchreuder) [reply docs/project/list-of-obsoletions.md]: Updated to use `SYSLIB0012`

File: src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/Hosting/AssemblyCatalogTests.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/Hosting/AssemblyCatalogTests.cs]: Updated to use `SYSLIB0012` as the pragma disable

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (JoshSchreuder) [reply src/libraries/System.Runtime/ref/System.Runtime.cs]: Removed `EditorBrowsableAttribute` per discussion

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs
Comment (JoshSchreuder) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs]: Do I need to obsolete this in a different way to support net462 etc. builds with the Diagnostic ID etc.?

Seems like adding the `[Obsolete]` here is breaking some builds, eg.
https://github.com/dotnet/runtime/pull/39261/checks?check_run_id=886299009

File: src/libraries/System.Runtime/ref/System.Runtime.csproj
Comment (jkotas) [inline src/libraries/System.Runtime/ref/System.Runtime.csproj:17]: The `ref/System.Runtime.cs` file should use the full literal string, not these constants. The file is auto re-generated from time to time and the auto generation would convert the constants into literal strings anyway.

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.csproj
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.csproj:8]: Same here.

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs
Comment (jkotas) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs]: It needs `#if NET50_OBSOLETIONS` in this file. Other places should be ok. They are not in components compiled for older runtime versions.

File: src/libraries/System.Runtime/ref/System.Runtime.csproj
Comment (JoshSchreuder) [reply src/libraries/System.Runtime/ref/System.Runtime.csproj:17]: @jkotas just to confirm, do you mean changing it to:

`<Compile Include="..\..\Common\src\System\Obsoletions.cs" Link="Common\System\Obsoletions.cs" />`? Or is there a better way to resolve the path than this?

Comment (jeffhandley) [reply src/libraries/System.Runtime/ref/System.Runtime.csproj]: He meant the messages and diagnostic is strings. Instead of pulling in `Obsoletions.cs` at all. Look for `SYSLIB003` in the ref .cs file to see the existing ones. We copy/paste the literal strings into the ref files.

Comment (JoshSchreuder) [reply src/libraries/System.Runtime/ref/System.Runtime.csproj]: Gotcha, will make that change, thanks

Comment (JoshSchreuder) [reply src/libraries/System.Runtime/ref/System.Runtime.csproj]: Updated in https://github.com/dotnet/runtime/pull/39261/commits/29f28769f25628bf532a382f6d7f9e89419e5413

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.csproj
Comment (JoshSchreuder) [reply src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.csproj]: Updated in https://github.com/dotnet/runtime/pull/39261/commits/29f28769f25628bf532a382f6d7f9e89419e5413

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jeffhandley) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:78]: I am not familiar with this code, so I need to defer to someone else to sign off on the changes to this method.

**Review by jeffhandley — COMMENTED** [review]
One tweak needed in a ref file; otherwise the obsoletion looks good. I am not familiar with the functional changes being made though, so we need to get another reviewer for those changes.

Thanks!

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:83]: It would be nice to cache `asm.Location` in a local to avoid calling it twice.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:86]:
```suggestion
                        string fileName = Path.GetFileNameWithoutExtension(asm.Location);
```

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:82]:
```suggestion
```

These comments do not look helpful.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:105]:
```suggestion
                else if (!resourceAssembly.IsDynamic)
```

I do not think it is necessary to check for empty location up-front here.

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs
Comment (JoshSchreuder) [reply src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:12]: @jeffhandley your one from #39269 didn't have this which I copied :D

https://github.com/dotnet/runtime/commit/329f0dbfa3a164ee2e3bf9f75b7641165995c799#diff-ef696a2876ba83b0d62cc5506022975c

Do you want me to update both?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:105]: @jkotas do you think it should be something more like

```csharp
else if (!resourceAssembly.IsDynamic)
{
    // Emitted ssemblies return empty string from Location.
    var location = asm.Location;
    if (location != string.Empty) 
    {
        string fileName = new FileInfo(location).Name;
        ...
    }
}
```

Comment (jkotas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:105]: I think it can be:
```csharp
else
{
    string location = asm.Location;
    if (location != string.Empty)
    {
        string fileName = Path.GetFileName(location);
...
```

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs
Comment (JoshSchreuder) [reply src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs]: Have updated both of these to use `System.ObsoleteAttribute`

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs]: Have updated

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs]: Removed comment

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs
Comment (JoshSchreuder) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs]: Have updated as suggested, thanks!

File: src/libraries/System.Runtime/ref/System.Runtime.csproj
Comment (JoshSchreuder) [reply src/libraries/System.Runtime/ref/System.Runtime.csproj]: > The file is auto re-generated from time to time and the auto generation would convert the constants into literal strings anyway.

Unrelated to this PR but I am curious, how / where does the process take place?

File: src/libraries/Common/src/System/Obsoletions.cs
Comment (jkotas) [inline src/libraries/Common/src/System/Obsoletions.cs:43]:
```suggestion
        internal const string CodeBaseMessage = "Use Location instead.";
```

The message approved by API-review is just "Use Location instead.": https://github.com/dotnet/runtime/issues/31127#issuecomment-545084027 . 

None of the other messages say that the APIs are only included for .NET Framework compatibility, even though number of them could.

File: docs/project/list-of-obsoletions.md
Comment (jkotas) [inline docs/project/list-of-obsoletions.md:27]:
```suggestion
|  __`SYSLIB0012`__ | Use Location instead. |
```

**Review by jkotas — APPROVED** [review]
Thank you!

File: src/libraries/System.Runtime/ref/System.Runtime.csproj
Comment (jkotas) [reply src/libraries/System.Runtime/ref/System.Runtime.csproj]: > Unrelated to this PR but I am curious, how / where does the process take place?

The documentation is at https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/updating-ref-source.md. The re-generation is done manually and produces many unwanted diffs today. #26187 tracks fixing it.

**Pull request merged**
